### PR TITLE
MIR unified ADTs

### PR DESCRIPTION
**Pull request by nagisa**
This PR changes translation of tuple-like ADTs from being calls to being proper aggregates. This change is done in hope to make code generation better. Namely, now we can avoid:
1. Call overhead;
2. Generating landingpads in presence of cleanups (we know for sure constructing ADTs can’t panic);
3. And probably much more, gaining better MIR introspectablilty.

Along with that a few serious deficiencies with translation of ADTs and switches have been fixed as well (commits 2 and 3).

r? @nikomatsakis 
cc @tsion

### CODE REVIEW

File: src/librustc_mir/hair/cx/expr.rs [hunk 1]
```diff
@@ -64,12 +63,40 @@ impl<'tcx> Mirror<'tcx> for &'tcx hir::Expr {
...
+                        ExprKind::Adt {
+                            adt_def: adt_def,
+                            substs: substs,
+                            variant_index: index,
+                            fields: field_refs,
+                            base: None
+                        }
+                    } else {
+                        ExprKind::Call {
+                            ty: &cx.tcx.node_id_to_type(fun.id),
```
Comment (tsion) [inline src/librustc_mir/hair/cx/expr.rs]: This was present before, but I think this & is redundant and will just get deref-coerced out of existence.

File: src/librustc_mir/hair/cx/expr.rs [hunk 2]
```diff
@@ -900,6 +929,7 @@ fn loop_label<'a, 'tcx: 'a>(cx: &mut Cx<'a, 'tcx>, expr: &'tcx hir::Expr) -> Cod
     }
 }
 
+/// Converts a list named (i.e. for struct-like struct/enum ADTs) into FieldExprRef.
```
Comment (tsion) [inline src/librustc_mir/hair/cx/expr.rs]: s/named/of named fields/

File: src/librustc_mir/hair/cx/expr.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/librustc_mir/hair/cx/expr.rs]: Yeah, probably a remnant of when the code was generic over the HIR.

**Comment by nikomatsakis** [comment]
Can you add a few more tests:

- using a tuple variable as a value (a `fn` value), e.g., `let x = Some; let y = x(3); y.unwrap()`
- struct tuple variants

Otherwise, looks good to me.

**Comment by nikomatsakis** [comment]
To be clear, r=me with more tests.

**Comment by nagisa** [comment]
@bors r=nikomatsakis 949dc72

(had omitted a rustc_mir attribute on test)

**Comment by arielb1** [comment]
We might want a more generic infrastructure to handle things like intrinsics, and maybe even trivial `Deref` functions.

**Comment by nagisa** [comment]
@bors r=nikomatsakis 72eacc1

(merge conflict with another patch which was not realised by git)

**Comment by nagisa** [comment]
@bors r=nikomatsakis 7d6da8e

(same reason, didn’t pay enough attention when dealing with the conflict)

**Pull request merged**